File: leetcode_python/2024.py
# 2024. 考试的最大困扰度
# https://leetcode-cn.com/problems/maximize-the-confusion-of-an-exam/
from typing import List
from collections import defaultdict
from collections import deque
from math import inf
import bisect
import heapq
'''
@File : 2024.py
@Time : 2022/03/11 23:18:36
@Author : YuMin Zhang
@State : Indepeedent | Half-Depedent | Depedent 
@Thinking :
@Tag : Easy | Medium | Hard
'''
class Solution:
    def maxConsecutiveAnswers(self, a: str, k: int) -> int:
        m = defaultdict(int)
        n, l, ans = len(a), 0, 0
        for r in range(n):
            m[a[r]] += 1
            if min(m["T"], m["F"]) > k:
                m[a[l]] -= 1
                l += 1
            ans = max(ans, r - l + 1)
        return ans

File: leetcode_c_plus_plus/2024.cpp
// 2024. 考试的最大困扰度
// https://leetcode-cn.com/problems/maximize-the-confusion-of-an-exam/
#include<map>
#include<set>
#include<vector>
#include<cmath>
#include<numeric>
#include<unordered_map>
using namespace std;
#define INT_MAX 2147483647
#define INT_MIN (-INT_MAX - 1)
/**
@File : 2024.cpp
@Time : 2022/03/29 20:39:56
@Author : YuMin Zhang
@State : Indepeedent | Half-Depedent | Depedent 
@Thinking :
@Tag : Easy | Medium | Hard
**/
class Solution {
public:
    int maxConsecutiveAnswers(string answerKey, int k) {
        int l = 0;
        int T = 0;
        int F = 0;
        int ans = 0;
        for(int i = 0; i < answerKey.size(); i++) {
            if(answerKey[i] == 'T') {
                T++;
            } else {
                F++;
            }
            while(min(T, F) > k) {
                if(answerKey[l] == 'T') T--;
                else F--;
                l++; 
            }
            ans = max(ans, i - l + 1);
        }
        return ans;
        
    }
};